### ISSUE TITLE
Issue in deserializing nullable SingleValueObjects

### ISSUE THREAD

**Issue by rutgervanwilligen** [eventflow/EventFlow#675]
Hi!

We have defined a `SingleValueObject<int>` in our domain and an event that contains a nullable version of this object. The resulting JSON of this event is either `{ "Key": 42 }` or `{ "Key": null }`.

Serializing to JSON works as expected, but deserializing is an issue because the deserializer does not recognize that the `SingleValueObject` itself is nullable (i.e. it tries to deserialize the value `null` into an `int`, which fails as `int` is a value type).

Let me start with a question: is adding a nullable `SingleValueObject` to an event a bad practice in the first place? I'd like to think it isn't (we've got valid use cases in our domain), but maybe there's something else we can/should do instead.

Second: could it be possible to change the `SingleValueObjectConverter` logic in such a way that it checks for a `null` value first, and deserializes only if a value of type `T` is found?

Related note: an issue might be that you can't distinguish anymore between single value objects that allow for `T = null` _within_ the object, and single value objects that don't. In our case, the fact that `T is int` (and not `int?`) implicates that it should be either instantiated with an int, or not at all -- a solution might be to add a `NullableSingleValueObject`, but it's a bit tedious.

I'd be happy to help with a PR, but like to know other people's thoughts on this first. Thanks in advance!

**Comment by rasmus** [comment]
Hi @rutgervanwilligen. It might that you have stumbled upon a bug. As you said, since your domain value can hold a null value, it should at least be `SingleValueObject<int?>`. I'll investigate

**Comment by rasmus** [comment]
Hi @rutgervanwilligen 

I did a few tests and are able to create values that are nullable as show below. Let me know if I misunderstood your question.

![image](https://user-images.githubusercontent.com/18956/64130271-93966500-cdc1-11e9-83c2-c3d0b666bae7.png)

**Comment by rasmus** [comment]
As show in https://github.com/eventflow/EventFlow/pull/677

**Comment by rutgervanwilligen** [comment]
Hi @rasmus, thanks for your reply! 

Interesting. I'll dive into my code again and (if it's still there) I'll try to extract it into its smallest reproducible form.

**Comment by rutgervanwilligen** [comment]
Hi @rasmus, I've seen your code. The difference between your test and our code is that in your test, the serialized JSON does not contain the key accompanying the `null` value. Here's a test that I would expect to succeed as well:

```
[Test]
public void DeserializeNullableIntWithKeyButWithoutValue()
{
    // Arrange
    var objectToSerialize = new WithNullableIntSingleValue(null);
    var json = JsonConvert.SerializeObject(objectToSerialize);

    // Act
    var with = JsonConvert.DeserializeObject<WithNullableIntSingleValue>(json);

    // Assert
    with.Should().NotBeNull();
    with.I.Should().BeNull();
}
```

But this throws a `JsonSerializationException` while trying to parse `null` to an `int`. Here, I would expect that the entire `SingleValueObject<int>` is deserialized as `null`. Instead, the serializer tries to deserialize it to an instantiated `SingleValueObject<int>` containing a `null` value.

**Comment by rutgervanwilligen** [comment]
> Hi @rutgervanwilligen. It might that you have stumbled upon a bug. As you said, since your domain value can hold a null value, it should at least be `SingleValueObject<int?>`. I'll investigate

Ah, I hadn't replied to this comment yet, sorry.

`SingleValueObject<int?>` is not allowed because of the `IComparable` constraint, but more importantly, I think it's conceptually different from what I'm trying to achieve. I believe that in our use case, the object we're defining as a `SingleValueObject<int>` should not be allowed instantiating with a nullable value, but rather "it should be either `null` as a whole, or instantiated with an `int`".

That way, our `null` checking logic moves one level higher; if we'd reason with this object, we could perform the null check on it and, if it's not `null`, retrieve its `Value`, instead of retrieving the `Value` blindly and performing the null check on the result.

(...does that make sense?)

**Comment by rasmus** [comment]
Got it, created a fix. Will release it asap

**Comment by rasmus** [comment]
Will do a release now

**Comment by rasmus** [comment]
Released in 0.75.3970

**Comment by rutgervanwilligen** [comment]
Awesome, thanks very much!
